common/pubsub/type_publisher.go

```diff
@@ -0,0 +1,124 @@
+package pubsub
+
+import (
+	"context"
+	"sync"
+
+	"github.com/status-im/status-go/common"
+)
+
+// Subscription represents a subscription to a publisher.
+type Subscription[T any] struct {
+	mu sync.RWMutex // protects the channel from being closed while a value is being published to it
+	ch chan T
+}
+
+func NewSubscription[T any]() *Subscription[T] {
+	return &Subscription[T]{
+		ch: make(chan T),
+	}
+}
+
+// Send the value val to the channel. If the context is canceled, the value is not sent.
+// The channel is protected from being closed while a value is being sent to it.
+func (s *Subscription[T]) Send(ctx context.Context, val T) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
```

Comment:
```suggestion
if ctx.Err() != nil {
    return
}
```

If both cases in the `select` are ready, Go chooses randomly. An early `ctx.Err()` check ensures no value is sent after cancellation.

Reply:
no context passed anymore since Publish is now non-blocking (events are dropped if the channel is busy)